doc/architecture/decisions/0011-use-redis-backed-drafts.md

```diff
@@ -0,0 +1,86 @@
+
+Also, the approach of embedding the data in the HTML is laborious. We have to make sure that every possible route through the pages in the journey preserves the data. This becomes particularly easy to get wrong when we have non-linear sequences of pages — for example, a link on the check your answers page that allows the user to edit a previous answer.
+
+### Other requirements for a solution
+
+The solution must also:
+
+- make sure that a user is not able to access data entered by a different user
+- not prevent the user from performing the same journey multiple times concurrently — for example, they should be able to assign two different interventions at the same time in different browser tabs
+- preserve the data that the user entered on previous pages when they use the browser’s Back button
+- give us maximum flexibility in deciding how to meet [WCAG 2.1’s Success Criterion 2.2.1 Timing Adjustable](https://www.w3.org/TR/WCAG21/#timing-adjustable) — for example, by making sure that the data is kept for at least 20 hours before it expires
+
+It would also be good if the solution could:
+
+- not introduce new dependencies to the service — for example a database
+- allow us to continue using the same kinds of coding patterns as we do when interacting with the interventions service API — creating and updating a resource
+- allow us to identify and clean up old data
+
+## Decision
+
+We’ll use the UI application’s existing Redis server as our storage. It allows us to store essentially unlimited amounts of data. The data is private to the current browsing session and is not accessible by other users.
+
+In Redis, we’ll store “draft” objects. These are containers of arbitrary JSON data, along with:
+
+- a globally unique identifier
+- the unique identifier of the user who created the draft
+- timestamps of creation and last update
+- an identifier explaining what type of data the draft represents
+
+We’ll:
+
+- provide a “CRUD” API for creating, fetching, updating and deleting these draft objects
+- include the draft’s ID in all URLs in a journey
+- allow these drafts to be created by a GET request, so that we can use `<a>` tags to link to journeys
+- prefer to use POSTs to pass data to a page instead of GET, so we don't have to worry about body size constraints
+
+The aforementioned API will:
+
+- enforce access control — making sure that a user is only allowed to access drafts that they created
+- make sure that drafts are automatically removed after they are no longer accessed for a certain amount of time (for example, 1 day), using Redis’s expiry functionality
+
+### Alternatives
+
+We might consider using the (Redis-backed) Express `session` object. However, this object expires after 2 hours of inactivity, which is insufficient for our needs. We don’t want to increase this timeout since there are security implications.
```

Comment:
Sorry, I might have missed this, but what is the timeout going to be in our approach? If longer than 2 hours, are there not security implications with our approach too?

Reply:
I think that the new timeout was implicit by the stuff about the WCAG expiry time requirement, and later on I say "make sure that drafts are automatically removed after they are no longer accessed for a certain amount of time (for example, 1 day)".

I'll add a bit explaining why a long-lived session is a security risk, but long-lived drafts aren't.

Reply:
New write-up makes sense, thanks for updating.